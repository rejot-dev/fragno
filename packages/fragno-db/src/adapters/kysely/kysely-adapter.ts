import { sql, type Kysely } from "kysely";
import type { SQLProvider } from "../../shared/providers";
import type { DatabaseAdapter } from "../adapters";
import { schemaToDBType } from "../../schema/serialize";
import { createMigrator, type Migrator } from "../../migration-engine/create";
import type { AnySchema } from "../../schema/create";
import type { CustomOperation, MigrationOperation } from "../../migration-engine/shared";
import { execute } from "./migration/execute";

const SETTINGS_TABLE_NAME = "fragno_db_settings" as const;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type KyselyAny = Kysely<any>;

export interface KyselyConfig {
  db: KyselyAny;
  provider: SQLProvider;
  namespace: string;
}

export class KyselyAdapter implements DatabaseAdapter {
  #kyselyConfig: KyselyConfig;

  constructor(config: KyselyConfig) {
    this.#kyselyConfig = config;
  }

  createMigrationEngine(schema: AnySchema): Migrator {
    const manager = createSettingsManager(
      this.#kyselyConfig.db,
      this.#kyselyConfig.provider,
      SETTINGS_TABLE_NAME,
    );

    const onCustomNode = (node: CustomOperation, db: KyselyAny) => {
      const statement = sql.raw(node["sql"] as string);

      return {
        compile() {
          return statement.compile(db);
        },
        execute() {
          return statement.execute(db);
        },
      };
    };

    const preprocess = (operations: MigrationOperation[], db: KyselyAny) => {
      if (this.#kyselyConfig.provider === "mysql") {
        operations.unshift({ type: "custom", sql: "SET FOREIGN_KEY_CHECKS = 0" });
        operations.push({ type: "custom", sql: "SET FOREIGN_KEY_CHECKS = 1" });
      } else if (this.#kyselyConfig.provider === "sqlite") {
        operations.unshift({
          type: "custom",
          sql: "PRAGMA defer_foreign_keys = ON",
        });
      }

      return operations.flatMap((op) =>
        execute(op, this.#kyselyConfig, (node) => onCustomNode(node, db)),
      );
    };

    return createMigrator({
      schema,
      userConfig: {
        provider: this.#kyselyConfig.provider,
      },
      executor: async (operations) => {
        await this.#kyselyConfig.db.transaction().execute(async (tx) => {
          for (const node of preprocess(operations, tx)) {
            try {
              await node.execute();
            } catch (e) {
              console.error("failed at", node.compile(), e);
              throw e;
            }
          }
        });
      },
      sql: {
        toSql: (operations) => {
          const compiled = preprocess(operations, this.#kyselyConfig.db).map(
            (m) => `${m.compile().sql};`,
          );

          return compiled.join("\n\n");
        },
      },

      settings: {
        getVersion: async () => {
          const v = await manager.get(`${this.#kyselyConfig.namespace}.schema_version`);
          return v ? parseInt(v) : 0;
        },
        updateSettingsInMigration: async (version) => {
          const init = await manager.initIfNeeded();

          if (init) {
            return [
              { type: "custom", sql: init },
              { type: "custom", sql: `UPDATE ${SETTINGS_TABLE_NAME} SET version = ${version}` },
            ];
          }

          return [
            { type: "custom", sql: `UPDATE ${SETTINGS_TABLE_NAME} SET version = ${version}` },
          ];
        },
      },
    });
  }

  getSchemaVersion() {
    const manager = createSettingsManager(
      this.#kyselyConfig.db,
      this.#kyselyConfig.provider,
      SETTINGS_TABLE_NAME,
    );

    return manager.get(`${this.#kyselyConfig.namespace}.schema_version`);
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function createSettingsManager(db: Kysely<any>, provider: SQLProvider, settingsTableName: string) {
  function initTable() {
    return db.schema
      .createTable(settingsTableName)
      .addColumn("key", provider === "sqlite" ? "text" : "varchar(255)", (col) => col.primaryKey())
      .addColumn("value", sql.raw(schemaToDBType({ type: "string" }, provider)), (col) =>
        col.notNull(),
      );
  }

  return {
    async get(key: string): Promise<string | undefined> {
      try {
        const result = await db
          .selectFrom(settingsTableName)
          .where("key", "=", key)
          .select(["value"])
          .executeTakeFirstOrThrow();
        return result.value as string;
      } catch {
        return;
      }
    },

    async initIfNeeded() {
      const tables = await db.introspection.getTables();
      if (tables.some((table) => table.name === settingsTableName)) {
        return;
      }

      return initTable().compile().sql;
    },

    insert(key: string, value: string) {
      return db
        .insertInto(settingsTableName)
        .values({
          key: sql.lit(key),
          value: sql.lit(value),
        })
        .compile().sql;
    },

    update(key: string, value: string) {
      return db
        .updateTable(settingsTableName)
        .set({
          value: sql.lit(value),
        })
        .where("key", "=", sql.lit(key))
        .compile().sql;
    },
  };
}
