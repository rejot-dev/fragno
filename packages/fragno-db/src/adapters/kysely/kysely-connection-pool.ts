import type { Kysely } from "kysely";
import type { ConnectionPool } from "../../shared/connection-pool";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type KyselyAny = Kysely<any>;

/**
 * Creates a Kysely-specific connection pool with proper cleanup.
 * Calls db.destroy() when the pool is closed to properly release database connections.
 */
export function createKyselyConnectionPool(
  dbOrFactory: KyselyAny | (() => KyselyAny | Promise<KyselyAny>),
): ConnectionPool<KyselyAny> {
  let cachedDb: KyselyAny | undefined;
  let initPromise: Promise<KyselyAny> | undefined;

  const ensureInitialized = async (): Promise<KyselyAny> => {
    if (cachedDb) {
      return cachedDb;
    }

    if (!initPromise) {
      initPromise = (async () => {
        const db =
          typeof dbOrFactory === "function"
            ? await (dbOrFactory as () => KyselyAny | Promise<KyselyAny>)()
            : dbOrFactory;
        cachedDb = db;
        return db;
      })();
    }

    return initPromise;
  };

  // Eagerly start initialization if it's a factory function
  if (typeof dbOrFactory === "function") {
    void ensureInitialized();
  } else {
    // Direct instance - cache it immediately
    cachedDb = dbOrFactory;
  }

  return {
    async connect() {
      const db = await ensureInitialized();

      return {
        db,
        release: async () => {},
      };
    },

    getDatabaseSync() {
      if (!cachedDb) {
        throw new Error("Cannot get database synchronously: database not initialized.");
      }
      return cachedDb;
    },

    async close() {
      if (cachedDb) {
        // Properly destroy Kysely instance to close all connections
        await cachedDb.destroy();
      }
      cachedDb = undefined;
      initPromise = undefined;
    },
  };
}
