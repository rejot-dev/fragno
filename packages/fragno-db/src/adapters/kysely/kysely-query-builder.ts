import {
  type BinaryOperator,
  type CompiledQuery,
  type ExpressionBuilder,
  type ExpressionWrapper,
  type Kysely,
  sql,
} from "kysely";
import type { AnySelectClause, FindManyOptions } from "../../query/query";
import type { SqlBool } from "kysely";
import { type AnyColumn, type AnyTable, Column } from "../../schema/create";
import type { SQLProvider } from "../../shared/providers";
import type { Condition } from "../../query/condition-builder";
import { serialize } from "../../schema/serialize";
import type { CompiledJoin, SimplifyFindOptions } from "../../query/orm/orm";
import { decodeResult, encodeValues, ReferenceSubquery } from "../../query/result-transform";

/**
 * Returns the fully qualified SQL name for a column (table.column).
 *
 * @param column - The column to get the full name for
 * @returns The fully qualified SQL name in the format "tableName.columnName"
 * @internal
 *
 * @example
 * ```ts
 * fullSQLName(userTable.columns.email)
 * // Returns: "users.email"
 * ```
 */
export function fullSQLName(column: AnyColumn) {
  return `${column.tableName}.${column.name}`;
}

/**
 * Builds a WHERE clause expression from a Condition tree.
 *
 * Recursively processes condition objects to build Kysely WHERE expressions.
 * Handles comparison operators, logical AND/OR/NOT, and special string operators
 * like "contains", "starts with", and "ends with".
 *
 * @param condition - The condition tree to build the WHERE clause from
 * @param eb - Kysely expression builder for constructing SQL expressions
 * @param provider - The SQL provider (affects SQL generation)
 * @returns A Kysely expression wrapper representing the WHERE clause
 * @internal
 *
 * @example
 * ```ts
 * const condition = {
 *   type: "compare",
 *   a: userTable.columns.name,
 *   operator: "contains",
 *   b: "john"
 * };
 * const whereClause = buildWhere(condition, eb, 'postgresql');
 * ```
 */
export function buildWhere(
  condition: Condition,
  eb: ExpressionBuilder<any, any>, // eslint-disable-line @typescript-eslint/no-explicit-any
  provider: SQLProvider,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
): ExpressionWrapper<any, any, SqlBool> {
  if (condition.type === "compare") {
    const left = condition.a;
    const op = condition.operator;
    let val = condition.b;

    if (!(val instanceof Column)) {
      val = serialize(val, left, provider);
    }

    let v: BinaryOperator;
    let rhs: unknown;

    switch (op) {
      case "contains":
        v = "like";
        rhs =
          val instanceof Column ? sql`concat('%', ${eb.ref(fullSQLName(val))}, '%')` : `%${val}%`;
        break;
      case "not contains":
        v = "not like";
        rhs =
          val instanceof Column ? sql`concat('%', ${eb.ref(fullSQLName(val))}, '%')` : `%${val}%`;
        break;
      case "starts with":
        v = "like";
        rhs = val instanceof Column ? sql`concat(${eb.ref(fullSQLName(val))}, '%')` : `${val}%`;
        break;
      case "not starts with":
        v = "not like";
        rhs = val instanceof Column ? sql`concat(${eb.ref(fullSQLName(val))}, '%')` : `${val}%`;
        break;
      case "ends with":
        v = "like";
        rhs = val instanceof Column ? sql`concat('%', ${eb.ref(fullSQLName(val))})` : `%${val}`;
        break;
      case "not ends with":
        v = "not like";
        rhs = val instanceof Column ? sql`concat('%', ${eb.ref(fullSQLName(val))})` : `%${val}`;
        break;
      default:
        v = op;
        rhs = val instanceof Column ? eb.ref(fullSQLName(val)) : val;
    }

    return eb(fullSQLName(left), v, rhs);
  }

  // Nested conditions
  if (condition.type === "and") {
    return eb.and(condition.items.map((v) => buildWhere(v, eb, provider)));
  }

  if (condition.type === "not") {
    return eb.not(buildWhere(condition.item, eb, provider));
  }

  return eb.or(condition.items.map((v) => buildWhere(v, eb, provider)));
}

/**
 * Maps a select clause to SQL column names with optional aliases.
 *
 * Converts application-level select clauses (either array of keys or "select all")
 * into SQL-compatible column selections with proper aliasing for relations.
 *
 * @param select - The select clause (array of keys or true for all columns)
 * @param table - The table schema containing column definitions
 * @param options - Optional configuration
 * @param options.relation - Relation name to prefix in aliases (for joined data)
 * @param options.tableName - Override the table name in the SQL (defaults to table.name)
 * @returns Array of SQL select strings in the format "tableName.columnName as alias"
 * @internal
 *
 * @example
 * ```ts
 * mapSelect(['id', 'name'], userTable)
 * // Returns: ['users.id as id', 'users.name as name']
 *
 * mapSelect(['title'], postTable, { relation: 'posts' })
 * // Returns: ['posts.title as posts:title']
 * ```
 */
export function mapSelect(
  select: AnySelectClause,
  table: AnyTable,
  options: {
    relation?: string;
    tableName?: string;
  } = {},
): string[] {
  const { relation, tableName = table.name } = options;
  const out: string[] = [];
  const keys = Array.isArray(select) ? select : Object.keys(table.columns);

  for (const key of keys) {
    const col = table.columns[key];

    // Skip hidden columns when explicitly selecting
    if (Array.isArray(select) && col.isHidden) {
      continue;
    }

    // Add the column to the select list
    const name = relation ? `${relation}:${key}` : key;
    out.push(`${tableName}.${col.name} as ${name}`);
  }

  // Always include hidden columns (for FragnoId construction with internal ID and version)
  for (const key in table.columns) {
    const col = table.columns[key];
    if (col.isHidden && !keys.includes(key)) {
      const name = relation ? `${relation}:${key}` : key;
      out.push(`${tableName}.${col.name} as ${name}`);
    }
  }

  return out;
}

/**
 * Result type from compiling a select clause with extensions.
 * @internal
 */
export interface CompiledSelect {
  /**
   * The final select clause to use in the query
   */
  result: AnySelectClause;

  /**
   * Keys that were added to the select clause (not originally requested)
   */
  extendedKeys: string[];

  /**
   * Removes the extended keys from a record (mutates the record).
   * Used to clean up keys that were only needed for join operations.
   *
   * @param record - The record to remove extended keys from
   * @returns The same record with extended keys removed
   */
  removeExtendedKeys: (record: Record<string, unknown>) => Record<string, unknown>;
}

/**
 * Builder for extending a select clause with additional keys.
 * @internal
 */
export interface SelectBuilder {
  /**
   * Adds a key to the select clause if not already present.
   * Tracks which keys were added for later removal.
   *
   * @param key - The key to add to the select clause
   */
  extend: (key: string) => void;

  /**
   * Compiles the select clause into its final form.
   *
   * @returns The compiled select information
   */
  compile: () => CompiledSelect;
}

/**
 * Creates a builder that can extend a select clause with additional keys.
 *
 * This is useful when you need to temporarily include columns for join operations
 * or other internal processing, but don't want them in the final result.
 *
 * @param original - The original select clause from the user
 * @returns A select builder with extend() and compile() methods
 * @internal
 *
 * @example
 * ```ts
 * const builder = extendSelect(['name', 'email']);
 * builder.extend('id'); // Add id for join operation
 * const { result, removeExtendedKeys } = builder.compile();
 * // result: ['name', 'email', 'id']
 *
 * const record = { name: 'John', email: 'j@ex.com', id: 123 };
 * removeExtendedKeys(record);
 * // record: { name: 'John', email: 'j@ex.com' }
 * ```
 */
export function extendSelect(original: AnySelectClause): SelectBuilder {
  const select = Array.isArray(original) ? new Set(original) : true;
  const extendedKeys: string[] = [];

  return {
    extend(key) {
      if (select === true || select.has(key)) {
        return;
      }

      select.add(key);
      extendedKeys.push(key);
    },
    compile() {
      return {
        result: select instanceof Set ? Array.from(select) : true,
        extendedKeys,
        removeExtendedKeys(record) {
          for (const key of extendedKeys) {
            delete record[key];
          }
          return record;
        },
      };
    },
  };
}

/**
 * Executes a SELECT query to find multiple records.
 *
 * Builds and executes a Kysely query with the provided options including
 * filtering (where), ordering (orderBy), pagination (limit/offset), and
 * column selection (select).
 *
 * @param kysely - The Kysely database instance
 * @param provider - The SQL provider (affects SQL generation)
 * @param table - The table to query from
 * @param v - Query options including where, select, orderBy, limit, and offset
 * @param runSubQueryJoin - Function to execute subquery joins on the results
 * @returns Array of decoded records matching the query criteria
 * @internal
 *
 * @example
 * ```ts
 * const records = await findMany(kysely, 'postgresql', userTable, {
 *   where: someCondition,
 *   orderBy: [['name', 'asc']],
 *   limit: 10
 * });
 * ```
 */
export async function findMany(
  kysely: Kysely<any>, // eslint-disable-line @typescript-eslint/no-explicit-any
  provider: SQLProvider,
  table: AnyTable,
  v: SimplifyFindOptions<FindManyOptions>,
  runSubQueryJoin: (records: Record<string, unknown>[], join: CompiledJoin) => Promise<void>,
) {
  let query = kysely.selectFrom(table.name);

  const where = v.where;
  if (where) {
    query = query.where((eb) => buildWhere(where, eb, provider));
  }

  if (v.offset !== undefined) {
    query = query.offset(v.offset);
  }

  if (v.limit !== undefined) {
    query = provider === "mssql" ? query.top(v.limit) : query.limit(v.limit);
  }

  if (v.orderBy) {
    for (const [col, mode] of v.orderBy) {
      query = query.orderBy(fullSQLName(col), mode);
    }
  }

  const selectBuilder = extendSelect(v.select);
  const mappedSelect: string[] = [];
  const subqueryJoins: CompiledJoin[] = [];

  const compiledSelect = selectBuilder.compile();
  mappedSelect.push(...mapSelect(compiledSelect.result, table));

  const records = (await query.select(mappedSelect).execute()).map((v) =>
    decodeResult(v, table, provider),
  );

  await Promise.all(subqueryJoins.map((join) => runSubQueryJoin(records, join)));
  for (const record of records) {
    compiledSelect.removeExtendedKeys(record);
  }

  return records;
}

/**
 * Processes encoded values and replaces ReferenceSubquery markers with actual SQL subqueries.
 *
 * @param values - The encoded values that may contain ReferenceSubquery objects
 * @param kysely - The Kysely database instance for building subqueries
 * @returns Processed values with subqueries in place of ReferenceSubquery markers
 * @internal
 */
function processReferenceSubqueries(
  values: Record<string, unknown>,
  kysely: Kysely<any>, // eslint-disable-line @typescript-eslint/no-explicit-any
): Record<string, unknown> {
  const processed: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(values)) {
    if (value instanceof ReferenceSubquery) {
      const refTable = value.referencedTable;
      const externalId = value.externalIdValue;

      // Build a subquery: SELECT _internal_id FROM referenced_table WHERE id = external_id LIMIT 1
      processed[key] = kysely
        .selectFrom(refTable.name)
        .select(refTable.getInternalIdColumn().name)
        .where(refTable.getIdColumn().name, "=", externalId)
        .limit(1);
    } else {
      processed[key] = value;
    }
  }

  return processed;
}

/**
 * Creates a query compiler that builds and compiles Kysely queries without executing them.
 *
 * Each method takes table and query parameters and returns a CompiledQuery that can be
 * executed later using kysely.executeQuery().
 *
 * @param kysely - The Kysely database instance
 * @param provider - The SQL provider (affects SQL generation)
 * @returns An object with methods for compiling various database operations
 * @internal
 *
 * @example
 * ```ts
 * const builder = createKyselyQueryBuilder(kysely, 'postgresql');
 * const query = builder.count(userTable, { where: someCondition });
 * const result = await kysely.executeQuery(query);
 * ```
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function createKyselyQueryBuilder(kysely: Kysely<any>, provider: SQLProvider) {
  return {
    count(table: AnyTable, { where }: { where?: Condition }): CompiledQuery {
      let query = kysely.selectFrom(table.name).select(kysely.fn.countAll().as("count"));
      if (where) {
        query = query.where((b) => buildWhere(where, b, provider));
      }
      return query.compile();
    },

    create(table: AnyTable, values: Record<string, unknown>): CompiledQuery {
      const encodedValues = encodeValues(values, table, true, provider);
      const processedValues = processReferenceSubqueries(encodedValues, kysely);
      const insert = kysely.insertInto(table.name).values(processedValues);

      if (provider === "mssql") {
        return (
          insert
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            .output(mapSelect(true, table, { tableName: "inserted" }) as any[])
            .compile()
        );
      }

      if (provider === "postgresql" || provider === "sqlite") {
        return insert.returning(mapSelect(true, table)).compile();
      }

      // For MySQL/other providers, return the insert query
      return insert.compile();
    },

    findMany<T extends AnyTable>(
      table: T,
      v: SimplifyFindOptions<FindManyOptions<T>>,
    ): CompiledQuery {
      let query = kysely.selectFrom(table.name);

      const where = v.where;
      if (where) {
        query = query.where((eb) => buildWhere(where, eb, provider));
      }

      if (v.offset !== undefined) {
        query = query.offset(v.offset);
      }

      if (v.limit !== undefined) {
        query = provider === "mssql" ? query.top(v.limit) : query.limit(v.limit);
      }

      if (v.orderBy) {
        for (const [col, mode] of v.orderBy) {
          query = query.orderBy(fullSQLName(col), mode);
        }
      }

      const selectBuilder = extendSelect(v.select);
      const mappedSelect: string[] = [];

      // Helper function to process joins recursively
      const processJoins = (
        joins: CompiledJoin[],
        parentTable: AnyTable,
        parentTableName: string,
        relationPrefix: string = "",
      ) => {
        for (const join of joins) {
          const { options: joinOptions, relation } = join;

          if (joinOptions === false) {
            continue;
          }

          const targetTable = relation.table;
          const joinName = relationPrefix ? `${relationPrefix}_${relation.name}` : relation.name;
          const fullRelationName = relationPrefix
            ? `${relationPrefix}:${relation.name}`
            : relation.name;

          // update select
          mappedSelect.push(
            ...mapSelect(joinOptions.select, targetTable, {
              relation: fullRelationName,
              tableName: joinName,
            }),
          );

          query = query.leftJoin(`${targetTable.name} as ${joinName}`, (b) =>
            b.on((eb) => {
              const conditions = [];
              for (const [left, right] of relation.on) {
                // Foreign keys always use internal IDs
                // If the relation references an external ID column (any name), translate to "_internalId"
                const rightCol = targetTable.columns[right];
                const actualRight = rightCol?.role === "external-id" ? "_internalId" : right;

                conditions.push(
                  eb(
                    `${parentTableName}.${parentTable.columns[left].name}`,
                    "=",
                    eb.ref(`${joinName}.${targetTable.columns[actualRight].name}`),
                  ),
                );
              }

              if (joinOptions.where) {
                conditions.push(buildWhere(joinOptions.where, eb, provider));
              }

              return eb.and(conditions);
            }),
          );

          // Process nested joins recursively
          if (joinOptions.join && joinOptions.join.length > 0) {
            processJoins(joinOptions.join, targetTable, joinName, fullRelationName);
          }
        }
      };

      // Process top-level joins
      if (v.join) {
        processJoins(v.join, table, table.name);
      }

      const compiledSelect = selectBuilder.compile();
      mappedSelect.push(...mapSelect(compiledSelect.result, table));

      return query.select(mappedSelect).compile();
    },

    updateMany(
      table: AnyTable,
      v: {
        where?: Condition;
        set: Record<string, unknown>;
      },
    ): CompiledQuery {
      const encoded = encodeValues(v.set, table, false, provider);
      const processed = processReferenceSubqueries(encoded, kysely);

      // Automatically increment _version for optimistic concurrency control
      const versionCol = table.getVersionColumn();
      // Safe cast: we're building a SQL expression for incrementing the version
      processed[versionCol.name] = sql.raw(`COALESCE(${versionCol.name}, 0) + 1`) as unknown;

      let query = kysely.updateTable(table.name).set(processed);
      const { where } = v;
      if (where) {
        query = query.where((eb) => buildWhere(where, eb, provider));
      }
      return query.compile();
    },

    upsertCheck(table: AnyTable, where: Condition | undefined): CompiledQuery {
      const idColumn = table.getIdColumn();
      let query = kysely.selectFrom(table.name).select([`${idColumn.name} as id`]);
      if (where) {
        query = query.where((b) => buildWhere(where, b, provider));
      }
      return query.limit(1).compile();
    },

    upsertUpdate(
      table: AnyTable,
      update: Record<string, unknown>,
      where: Condition | undefined,
      top?: boolean,
    ): CompiledQuery {
      const encoded = encodeValues(update, table, false, provider);
      const processed = processReferenceSubqueries(encoded, kysely);
      let query = kysely.updateTable(table.name).set(processed);
      if (top) {
        query = query.top(1);
      }
      if (where) {
        query = query.where((b) => buildWhere(where, b, provider));
      }
      return query.compile();
    },

    upsertUpdateById(table: AnyTable, update: Record<string, unknown>, id: unknown): CompiledQuery {
      const idColumn = table.getIdColumn();
      const encoded = encodeValues(update, table, false, provider);
      const processed = processReferenceSubqueries(encoded, kysely);
      return kysely.updateTable(table.name).set(processed).where(idColumn.name, "=", id).compile();
    },

    createMany(table: AnyTable, values: Record<string, unknown>[]): CompiledQuery {
      const encodedValues = values.map((v) => encodeValues(v, table, true, provider));
      const processedValues = encodedValues.map((v) => processReferenceSubqueries(v, kysely));
      return kysely.insertInto(table.name).values(processedValues).compile();
    },

    deleteMany(table: AnyTable, { where }: { where?: Condition }): CompiledQuery {
      let query = kysely.deleteFrom(table.name);
      if (where) {
        query = query.where((eb) => buildWhere(where, eb, provider));
      }
      return query.compile();
    },

    findById(table: AnyTable, idValue: unknown): CompiledQuery {
      const idColumn = table.getIdColumn();
      return kysely
        .selectFrom(table.name)
        .select(mapSelect(true, table))
        .where(idColumn.name, "=", idValue)
        .limit(1)
        .compile();
    },
  };
}
