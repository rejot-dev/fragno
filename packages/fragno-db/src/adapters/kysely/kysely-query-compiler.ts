import type { CompiledQuery } from "kysely";
import type { AbstractQueryCompiler } from "../../query/query-compiler";
import type { AnySchema, AnyTable } from "../../schema/create";
import { buildCondition } from "../../query/condition-builder";
import { buildFindOptions } from "../../query/orm/orm";
import type { KyselyConfig } from "./kysely-adapter";
import { createKyselyQueryBuilder } from "./query-builder";
import { encodeValues } from "./result-transform";

export function createKyselyQueryCompiler<T extends AnySchema>(
  schema: T,
  config: KyselyConfig,
): AbstractQueryCompiler<T, CompiledQuery> {
  const { db: kysely, provider } = config;
  const queryBuilder = createKyselyQueryBuilder(kysely, provider);

  function toTable(name: unknown): AnyTable {
    const table = schema.tables[name as string];
    if (!table) {
      throw new Error(`Invalid table name ${name}.`);
    }
    return table;
  }

  return {
    count(name, { where } = {}) {
      const table = toTable(name);
      let conditions = where ? buildCondition(table.columns, where) : undefined;
      if (conditions === true) {
        conditions = undefined;
      }
      if (conditions === false) {
        return null;
      }

      return queryBuilder.count(table, { where: conditions });
    },

    findFirst(name, options) {
      const table = toTable(name);
      // Safe cast: FindFirstOptions is structurally compatible with FindManyOptions
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const compiledOptions = buildFindOptions(table, options as any);
      if (compiledOptions === false) {
        return null;
      }

      return queryBuilder.findMany(table, {
        ...compiledOptions,
        limit: 1,
      });
    },

    findMany(name, options = {}) {
      const table = toTable(name);
      // Safe cast: FindManyOptions from compiler matches FindManyOptions from buildFindOptions
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const compiledOptions = buildFindOptions(table, options as any);
      if (compiledOptions === false) {
        return null;
      }

      return queryBuilder.findMany(table, compiledOptions);
    },

    create(name, values) {
      const table = toTable(name);
      // Encode values before passing to query builder
      const encodedValues = encodeValues(values, table, true, provider);

      return queryBuilder.create(table, encodedValues);
    },

    createMany(name, values) {
      const table = toTable(name);
      // Encode all values before passing to query builder
      const encodedValues = values.map((v) => encodeValues(v, table, true, provider));

      return queryBuilder.createMany(table, encodedValues);
    },

    updateMany(name, { set, where }) {
      const table = toTable(name);
      let conditions = where ? buildCondition(table.columns, where) : undefined;
      if (conditions === true) {
        conditions = undefined;
      }
      if (conditions === false) {
        return null;
      }

      // Encode the set values
      const encodedSet = encodeValues(set, table, false, provider);

      return queryBuilder.updateMany(table, { set: encodedSet, where: conditions });
    },

    deleteMany(name, { where }) {
      const table = toTable(name);
      let conditions = where ? buildCondition(table.columns, where) : undefined;
      if (conditions === true) {
        conditions = undefined;
      }
      if (conditions === false) {
        return null;
      }

      return queryBuilder.deleteMany(table, { where: conditions });
    },
  } satisfies AbstractQueryCompiler<T, CompiledQuery>;
}
