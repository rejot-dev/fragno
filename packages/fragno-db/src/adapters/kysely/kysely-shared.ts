/**
 * Maps logical table names (used by fragment authors) to physical table names (with namespace suffix)
 */
export interface TableNameMapper {
  toPhysical(logicalName: string): string;
  toLogical(physicalName: string): string;
}

/**
 * Creates a table name mapper for a given namespace.
 * Physical names have format: {logicalName}_{namespace}
 */
export function createTableNameMapper(namespace: string): TableNameMapper {
  return {
    toPhysical: (logicalName: string) => `${logicalName}_${namespace}`,
    toLogical: (physicalName: string) => {
      if (physicalName.endsWith(`_${namespace}`)) {
        return physicalName.slice(0, -(namespace.length + 1));
      }
      return physicalName;
    },
  };
}
