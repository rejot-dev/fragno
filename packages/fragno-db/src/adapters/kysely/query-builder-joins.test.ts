import { Kysely } from "kysely";
import { KyselyPGlite } from "kysely-pglite";
import { describe, it, beforeAll, assert, expect } from "vitest";
import { column, idColumn, referenceColumn, schema } from "../../schema/create";
import type { Kysely as KyselyType } from "kysely";
import { createKyselyQueryCompiler } from "./kysely-query-compiler";

describe("query-builder-joins", () => {
  const userSchema = schema((s) => {
    return s
      .addTable("users", (t) => {
        return t.addColumn("id", idColumn()).addColumn("name", column("string"));
      })
      .addTable("posts", (t) => {
        return t
          .addColumn("id", idColumn())
          .addColumn("title", column("string"))
          .addColumn("userId", referenceColumn());
      })
      .addTable("tags", (t) => {
        return t.addColumn("id", idColumn()).addColumn("name", column("string"));
      })
      .addReference("posts", "author", {
        columns: ["userId"],
        targetTable: "users",
        targetColumns: ["id"],
      });
  });

  let kysely: KyselyType<any>; // eslint-disable-line @typescript-eslint/no-explicit-any

  beforeAll(async () => {
    const { dialect } = await KyselyPGlite.create();
    kysely = new Kysely({ dialect });
  });

  describe("postgresql", () => {
    it("should compile select with join condition comparing columns", () => {
      const compiler = createKyselyQueryCompiler(userSchema, {
        db: kysely,
        provider: "postgresql",
      });

      const query = compiler.findMany("posts", {
        select: ["id", "userId"],
        join: (b) => b.author(),
      });

      assert(query);
      expect(query.sql).toMatchSnapshot();
    });
  });
});
