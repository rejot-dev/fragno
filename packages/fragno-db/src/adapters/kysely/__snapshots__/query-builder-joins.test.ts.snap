// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`query-builder-joins > postgresql > should compile select with join condition comparing columns 1`] = `"select "author"."id" as "author:id", "author"."name" as "author:name", "posts"."id" as "id", "posts"."userId" as "userId" from "posts" left join "users" as "author" on "posts"."userId" = "author"."id""`;
